### ISSUE TITLE
Aim for the first working version

### ISSUE THREAD

**Issue by GeorgeS2019** [Remi123/MotionMatching#5]
- [ ] Demo or any explanation on how to setup it up.
- [ ] Animations with root motion ( you have to provide them yourself.)
- [ ] Character.
- [ ] Inertialization or any method to blend the current animation to the result of the motionmatcher query.
- [ ]  Get the plugin working by docking the bottom Panel with **MOCKUP** data
![image](https://github.com/Remi123/MotionMatching/assets/49812372/4151c570-698c-4aac-8e84-d8fa8066ef52)

**Comment by GeorgeS2019** [comment]
@CBerry22

On behalf of many of us who care about the Godot community, I wonder if you could apply your experience with [physics-based active ragdoll to motion matching](https://github.com/CBerry22/Active-Ragdoll---Physics-Animations-in-Godot-4.0)

There are similarities between them
- Motion Matching
> **Inertialization** or any method to blend the **current animation** to the result of the **motionmatcher query.**

- Active ragdoll 
> Some sort of blend to adjust the physic-based character to the result of query the animation

# To learn about Motion Matching
 (provided by Readme)

[Introduction video](https://www.gdcvault.com/play/1023280/Motion-Matching-and-The-Road)
[O3DE implementation blog](https://github.com/o3de/o3de/tree/development/Gems/MotionMatching)
[Daniel Holden aka OrangeDuck's blog](https://theorangeduck.com/)
[Spring algorithm and controller](https://theorangeduck.com/page/spring-roll-call)
[Motion Matching implementation](https://theorangeduck.com/page/code-vs-data-driven-displacement)
[Fitting Motion with Animation](https://theorangeduck.com/page/fitting-code-driven-displacement)

# Discussion and elaboration provided by @Remi123 

This project has scalability as one of its core foundation. Since many users will try to use it with very different quality of animation, from professional post-processed motion-captured animations to Mixamo animations to their own animations of varying quality, the system must be robust, easy to debug and easy to fix.

## Nomenclature:

- **Features** : Extract information, most of them time in the model local space relative to the root bone. Example include current velocity, bone positions and velocities, and trajectory point at x point in the future. Each features has a dimension size required for serializing the data. The sum of all the features dimensions is named K, because we will use a KDTree
- Serialization : In MM term, we populate all the features information into an 1D array of size K.
- Baking : The act of looking at each animations and extracting the features information at time T and adding those information to the KDTree.

Some comments about my project and some possibilities it allows.

1. Animation Re-targeting make importing BVH, FBX and GLTF animation a non-issues for this project. My own project use a mix of motion captured and mixamo animations without trouble.
2. Debugging is important. Each features is able to show a gizmo ( only in the editor for the moment) showing lines or meshes to show velocity, bone position, etc.
3. The Kdtree implementation that we choose has three distinct advantages over using other methods : Weight, void* Data and a Predicate,
    - Weight allows to specify the importance of different features over others. The trajectory is usually the most important in locomotion.
    - Void* allows me to reference some data at each node. I use it to store a bitfield representing the category ( Walk, run, etc)
    - Predicate allows to filter results. I use it for the categories, but I also have ideas in the editor.

What this project aims is to provide multiple type of features, ones more adapted for typical locomotion that are continuously updated, others would be better for Melee Combat similar to For Honor, or other better used for Parkour. Each would be hosted by it's own kdtree and the user can select which one are active by code. This is what we define as scalability in this project.

This project is still a WIP and as such some features were implemented as a prototype for now.

**Comment by GeorgeS2019** [comment]
@maximkulkin

With respect to your feedback on [**AminationTreeNode with parameters**](https://github.com/godotengine/godot-proposals/issues/3352#issuecomment-1510282834), I think this project needs to address this possibility too. value your input if you have time.

**Comment by GeorgeS2019** [comment]
@SaracenOne

[Retool AnimationTree parameter system to use a blackboard-based approach](https://github.com/godotengine/godot-proposals/issues/3352)

In this project, we are also facing how to deal with **AinmationTree parameter system** to a blackboard that returns a motion-matched query. Value your feedback

**Comment by Remi123** [comment]
@GeorgeS2019 @CBerry22 @maximkulkin @SaracenOne

I can share my inputs on each subject.

### AnimationTree

My plugin is outside AnimationTree for now, it's only a node that find the best poses and output the name and timestamp of the animation that best match it. For example, "WalkForward" at time 0.4. This is all my plugin output, but I do it every 0.2 seconds.
 
Once this is known I need to blend to the desired pose. I have a simple setup in the animationtree that blend between two poses, but basically all I need is a node that blend between the current pose, and the poses at my desired animation timestamp. 

I value the fact that my setup is technically just a slightly more complex AnimationNode, as the user can use all the things that AnimationTree is able to output.

### Inertialization vs active ragdoll.

I remembered that he had a Youtube channel, and looked back at the active ragdoll at [this link](https://www.youtube.com/watch?v=0MHY2TDeMLM&t=304s). Quite frankly I'm impressed and I think it's possible to implement the blending using active ragdoll, but I have some reservation.

### Issues 1
First of all, he is using two skeletons, one physical using PhysicalBone3D and the other that is playing the animation. The physical skeleton is trying to match the position of the equivalent bone in the normal rig using some spring calculation. It is indeed very similar to inertialization which is also derived from springs, although the calculation isn't the same but for this discussion it's similar enough.

With my plugin, the animated rig would simply go to the desired poses, wait 0.X seconds that is configurable, and continue playing the rest of the animation and the physical bones would simply catch up with the rest of the animation.

I'll try to make it so that there is only one skeleton, but I have to test it if is possible.

### Issues 2 
For my purpose, it is still a workaround a proper node that blend realistically between poses very often. My setup that switch between the current and desired poses would just be replaced by an active ragdoll setup that still needs maintenance, but now with double the skeletons.

### Inertialization

I've discussed with Tokage about a solution to the inertialization. In essence, the real problem that I have to workaround is that I cannot create an AnimationNode that manipulate the Skeleton3D directly. The animationtree has it's own structure to represent a blend of all the nodes and I want to keep it that way.

My current solution would be to have a node similar to an AnimationNode, and once it receive the signal to switch to a desired pose, create a temporary animation with the current bones positions and rotations. I then smoothly transitions to the desired pose by manipulating the animations tracks's key values instead of the skeleton bones positions. The animationTree only see an Animation and I'm free to manipulate the values however I see fit. Once the transition is done I simply remove the temporary animation and play the desired one.

**Comment by GeorgeS2019** [comment]
@Remi123 

Your **approach** is an outlier compared to almost the majority of approaches taken by Unity communities. I have not evaluated the Unreal community.

Your approach is VERY **UN-SpaceX**, lack public agile iteration from simple to your complex "outlier" approach

**Comment by GeorgeS2019** [comment]
@Remi123 
I just found out about significant progress in another branch.

When is the first working demo, not just the addon?

**Comment by Remi123** [comment]
@GeorgeS2019 

I've realized that I had lot of issues coming from my initial setup in my project, so I've started a new one and fixed most of the them. Basically stuff like skeletons's nodepath to their bones and bla bla bla. A bunch of bugs were fixed also. But here is the latest update : 

- Inertialization

The biggest challenge was inertialization, or the blend between poses. I happy to say I finally have it, but I wasn't able to make it part of an AnimationNode and integrate it with the AnimationTree. I've tried various approach, even creating per-skeleton animation and modifying the keys at runtime but it started messing with the cache of animationtree and the skeleton wasn't updating.

After talking with @TokageitLab and @fire, I was made aware that AnimationTree and AnimationPlayer are going throught a big refactoring, and since I was already trying to squeeze in this feature, I made the decision to delay implementing it in way that is compatible with the current setup, and just provide a class that inherit Skeleton3D and do the inertialization. I'll show the class to the implementers of AnimationTree so that they can decide how to integrate it if they want.

In the meantime there is an animationtree setup that can blend between poses but isn't as good as inertialization. I can include it in the demo and talk about it in the wiki, but I'll continue the development using my own Skeleton3D class.

Here a video of a inertialization test. Basically it's switching which animation it run every second,  from a random time in the animation. The cool thing is that I can switch animation mid-switch. Still WIP ( first second isn't set I'll fix it later, and it float due to the skeleton weird rest pose.)
[july30.webm](https://github.com/Remi123/MotionMatching/assets/16803994/ac1c789e-9ff3-44bd-8fd1-c52003c75bea)

- Demo

A demo, while important, doesn't explain everything so it is more important to have a good wiki. I've started writing it but of course I've just started.

- Root Motion Animation

For now I'm using modified Mixamo animations with root motion analyze and play. Of course I can't distribute them so I made the whole plugin able to analyze any animation in an AnimationLibrary. I've talked with Fire and they have a setup to capture and record animations so my goal is to write a document saying what would be good for Motion Matching locomotion. If they want to share it it's great, the demo can include it but like I said before any animation can be used.

- Character.

I've just used Kenney's character since they are free and CC0. With animation retargeting it become a non issues, which is already implemented in Godot. In fact my inertialization only look into the animation data, and only apply the motion_scale at the end.

- Get the plugin working by docking the bottom Panel with MOCKUP data

I'm not exactly sure what you mean with this. 

- Release of the demo.
Not sure when. I have a few things to fix as the inertialization is just finally working.
